### PR TITLE
Pytorch autolog Examples - Fixing argument parser & Adding `accelerator` to MLProject

### PR DESCRIPTION
**Pull request by shrinath-suresh**
Signed-off-by: Shrinath Suresh <shrinath@ideas2it.com>

## What changes are proposed in this pull request?

Breaking Change - Number of epochs to run argument has been set to `max_epochs` in [README.md](https://github.com/mlflow/mlflow/blob/master/examples/pytorch/MNIST/example1/README.md) and [MLProject](https://github.com/mlflow/mlflow/blob/master/examples/pytorch/MNIST/example1/MLproject). However, all the [examples](https://github.com/mlflow/mlflow/blob/c56d92002fc6dae387e54684b6a61f25896edc9f/examples/pytorch/MNIST/example1/mnist_autolog_example1.py#L281) are still set as `max-epochs` .

Hence, Changing it to `max_epochs` in all the scripts.

Adding `accelerator` parameter to MLProject. For CPU users accelerator parameter will be None by default.

## How is this patch tested?

Tested by running the examples

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [x] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by shrinath-suresh** [comment]
@smurching @chauhang I have set the max epoch parameter to `max_epochs`. 

Having the arguments as `_` in one place and `-` in another is super confusing. Can we change all the Pytorch Lightning arguments to `_` . Ex: `batch_size` . Since, the MLProject variables are already having `_` in it. It will look uniform and user friendly. Your thoughts ?

**Review by chauhang — APPROVED** [review]

**Comment by SeanNaren** [comment]
What do we think about using:

```python
import pytorch_lightning as pl

parser  = pl.Trainer.add_argparse_args(parent_parser=parser)
args = parser.parse_args()
```
This adds all the pl defaults args without having to be explicit. Then we can remove all the trainer related args?

File: examples/pytorch/MNIST/example1/mnist_autolog_example1.py [hunk 1]
```diff
@@ -278,13 +278,16 @@ def configure_optimizers(self):
     # Add trainer specific arguments
 
     parser.add_argument(
-        "--max-epochs", type=int, default=20, help="number of epochs to run (default: 20)"
+        "--max_epochs", type=int, default=20, help="number of epochs to run (default: 20)"
```
Comment (harupy) [inline examples/pytorch/MNIST/example1/mnist_autolog_example1.py:281]: @shrinath-suresh Can we revert this change?

Comment (shrinath-suresh) [reply examples/pytorch/MNIST/example1/mnist_autolog_example1.py:281]: @harupy This is the breaking change. The number of epochs parameter is set to `max_epochs` in read me files and MLProject and still it was  `max-epochs` in script.

Comment (harupy) [reply examples/pytorch/MNIST/example1/mnist_autolog_example1.py:281]: Can we fix README instead?

**Comment by shrinath-suresh** [comment]
> What do we think about using:
> 
> ```python
> import pytorch_lightning as pl
> 
> if __name__ == '__main__':
>     ...
>     parser  = pl.Trainer.add_argparse_args(parent_parser=parser)
>     args = parser.parse_args()
> ```
> 
> This adds all the pl defaults args without having to be explicit. Then we can remove all the trainer related args?

Thanks for the suggestion, i will test this one with MLProject and let you know

File: examples/pytorch/MNIST/example1/mnist_autolog_example1.py (hunk 1 above)
Comment (chauhang) [reply examples/pytorch/MNIST/example1/mnist_autolog_example1.py:281]: @harupy The name changes were done per previous request from @smurching Can you please check with him on what we want as the final naming scheme. For `mlflow run .` the underscore and hypens are not interchangeable

Comment (smurching) [reply examples/pytorch/MNIST/example1/mnist_autolog_example1.py]: Sorry for the confusion, I think it's fine to be consistent in our usage of ``--max-epochs`` across the script and MLproject file. My point in the other PR was more that we should be consistent in naming a parameter `max-epochs` in both places rather than `epochs` in one and `max-epochs` elsewhere, I didn't mean to express a preference in favor of underscores

In general, the POSIX convention is to use hyphens (`-`) rather than underscores (`_`) in parameter names: https://www.gnu.org/software/libc/manual/html_node/Argument-Syntax.html, so seems reasonable to follow the same convention here, and use hyphens in both the script & MLproject file as @harupy is suggesting

**Review by SeanNaren — APPROVED** [review]

File: examples/pytorch/MNIST/example1/mnist_autolog_example1.py (hunk 1 above)
Comment (shrinath-suresh) [reply examples/pytorch/MNIST/example1/mnist_autolog_example1.py]: @smurching @harupy @chauhang Pytorch Lightning uses `_` in argument parser (Ex: `max_epochs`). As @SeanNaren suggested in https://github.com/mlflow/mlflow/pull/3655#issuecomment-724222640 I have removed all the trainer specific arguments and used PTL's argument options. Hence, we cannot pass `max-epochs`anymore with the current code. 

To maintain the consistency, i will update the variables to have `_` (Ex: `batch_size` , `num_workers`).

Comment (harupy) [reply examples/pytorch/MNIST/example1/mnist_autolog_example1.py]: @shrinath-suresh 

> we cannot pass `max-epochs` anymore with the current code.

Does this mean 

```
python xxx.py --max-epochs 3
```

Comment (shrinath-suresh) [reply examples/pytorch/MNIST/example1/mnist_autolog_example1.py]: Yes you are right 

```
python bert_classification.py --max-epochs 1

usage: bert_classification.py [-h] [--num-samples N] [--logger [LOGGER]] [--checkpoint_callback [CHECKPOINT_CALLBACK]] [--default_root_dir DEFAULT_ROOT_DIR] [--gradient_clip_val GRADIENT_CLIP_VAL]
                              [--process_position PROCESS_POSITION] [--num_nodes NUM_NODES] [--num_processes NUM_PROCESSES] [--gpus GPUS] [--auto_select_gpus [AUTO_SELECT_GPUS]] [--tpu_cores TPU_CORES]
                              [--log_gpu_memory LOG_GPU_MEMORY] [--progress_bar_refresh_rate PROGRESS_BAR_REFRESH_RATE] [--overfit_batches OVERFIT_BATCHES] [--track_grad_norm TRACK_GRAD_NORM]
                              [--check_val_every_n_epoch CHECK_VAL_EVERY_N_EPOCH] [--fast_dev_run [FAST_DEV_RUN]] [--accumulate_grad_batches ACCUMULATE_GRAD_BATCHES] [--max_epochs MAX_EPOCHS]
                              [--min_epochs MIN_EPOCHS] [--max_steps MAX_STEPS] [--min_steps MIN_STEPS] [--limit_train_batches LIMIT_TRAIN_BATCHES] [--limit_val_batches LIMIT_VAL_BATCHES]
                              [--limit_test_batches LIMIT_TEST_BATCHES] [--val_check_interval VAL_CHECK_INTERVAL] [--flush_logs_every_n_steps FLUSH_LOGS_EVERY_N_STEPS]
                              [--log_every_n_steps LOG_EVERY_N_STEPS] [--accelerator ACCELERATOR] [--sync_batchnorm [SYNC_BATCHNORM]] [--precision PRECISION] [--weights_summary WEIGHTS_SUMMARY]
                              [--weights_save_path WEIGHTS_SAVE_PATH] [--num_sanity_val_steps NUM_SANITY_VAL_STEPS] [--truncated_bptt_steps TRUNCATED_BPTT_STEPS]
                              [--resume_from_checkpoint RESUME_FROM_CHECKPOINT] [--profiler [PROFILER]] [--benchmark [BENCHMARK]] [--deterministic [DETERMINISTIC]]
                              [--reload_dataloaders_every_epoch [RELOAD_DATALOADERS_EVERY_EPOCH]] [--auto_lr_find [AUTO_LR_FIND]] [--replace_sampler_ddp [REPLACE_SAMPLER_DDP]]
                              [--terminate_on_nan [TERMINATE_ON_NAN]] [--auto_scale_batch_size [AUTO_SCALE_BATCH_SIZE]] [--prepare_data_per_node [PREPARE_DATA_PER_NODE]] [--amp_backend AMP_BACKEND]
                              [--amp_level AMP_LEVEL] [--distributed_backend DISTRIBUTED_BACKEND] [--automatic_optimization [AUTOMATIC_OPTIMIZATION]] [--lr LR] [--batch-size N] [--num-workers N]
bert_classification.py: error: unrecognized arguments: --max-epochs 1

```

Comment (harupy) [reply examples/pytorch/MNIST/example1/mnist_autolog_example1.py]: do we really need these options?

Comment (shrinath-suresh) [reply examples/pytorch/MNIST/example1/mnist_autolog_example1.py]: The big list is from Pytorch Lightning's argument list. Not from the script.

Comment (chauhang) [reply examples/pytorch/MNIST/example1/mnist_autolog_example1.py]: @harupy In the MLProject files, we have kept the relevant parameters for ease of reproducibility. All the others mentioned above are the optional prams that PTL supports

**Review by harupy — APPROVED** [review]
LGTM
